src/services/blocks/trace/Trace.ts

```diff
@@ -0,0 +1,666 @@
+ *
+ * @param api ApiPromise
+ */
+function getKeyNames(api: ApiPromise): Record<string, KeyInfo> {
+	const moduleKeys = Object.keys(api.query).reduce((acc, mod) => {
+		Object.keys(api.query[mod]).forEach((item) => {
+			const queryObj = api.query[mod][item];
+			let key;
+			try {
+				// Slice off '0x' prefix
+				key = queryObj.key().slice(2);
+			} catch {
+				key = queryObj.keyPrefix().slice(2);
+			}
+
+			acc[key] = {
+				module: mod,
+				item,
+			};
+		});
+
+		return acc;
+	}, {} as Record<string, KeyInfo>);
+
+	const wellKnownKeys = {
+		'3a636f6465': {
+			special: ':code',
+		},
+		'3a686561707061676573': {
+			special: ':heappages',
+		},
+		'3a65787472696e7369635f696e646578': {
+			special: ':extrinsic_index',
+		},
+		'3a6368616e6765735f74726965': {
+			special: ':changes_trie',
+		},
+		'3a6368696c645f73746f726167653a': {
+			special: ':child_storage:',
+		},
+	};
+
+	return { ...moduleKeys, ...wellKnownKeys };
+}
+
+/**
+ * Parses spans by
+ * 1) creating a Map of span's `id` to the the span with its children
+ * 2) creating an array with the same spans
+ *
+ * @param spans spans to parse
+ */
+function parseSpans(
+	spans: TraceSpan[]
+): {
+	spansById: Map<number, SpanWithChildren>;
+	parsedSpans: SpanWithChildren[];
+	executeBlockSpanId: number;
+} {
+	// IMPORTANT for downstream logic: Make sure spans are sorted.
+	spans.sort((a, b) => a.id - b.id);
+
+	const spansById = new Map<number, SpanWithChildren>();
+	const parsedSpans = [];
+	let executeBlockSpanId;
+	for (const span of spans) {
+		const spanWithChildren = { ...span, children: [] };
+
+		if (
+			span.name === SPANS.executeBlock.name &&
+			span.target === SPANS.executeBlock.target
+		) {
+			executeBlockSpanId = span.id;
+		}
+
+		parsedSpans.push(spanWithChildren);
+		spansById.set(span.id, spanWithChildren);
+	}
+
+	if (executeBlockSpanId === undefined) {
+		throw new InternalServerError('execute_block span could not be found');
+	}
+
+	// Go through each span, and add its Id to the `children` array of its
+	// parent span, creating tree of spans starting with `execute_block` span as
+	// the root.
+	for (const span of spans) {
+		if (!span.parentId) {
+			continue;
+		}
+
+		if (!spansById.has(span.parentId)) {
+			throw new InternalServerError(
+				'Expected spans parent to exist in spansById'
+			);
+		}
+
+		spansById.get(span.parentId)?.children.push(span.id);
+	}
+
+	return { spansById, parsedSpans, executeBlockSpanId };
+}
+
+/**
+ * Find the Ids of al the descendant spans of `root`.
+ *
+ * @param root span which we want all the descendants of
+ * @param spansById map of span id => span with children
+ */
+function findDescendants(
+	root: number,
+	spansById: Map<number, SpanWithChildren>
+): number[] {
+	// Note: traversal order doesn't matter here
+	const stack = spansById.get(root)?.children;
+	if (!stack) {
+		return [];
+	}
+
+	const descendants = [];
+	while (stack.length) {
+		const curId = stack.pop();
+		if (!curId) {
```

Comment:
`if (!curId) break;` ?

Reply:
I prefer multi line with brackets for consistency and readability.